### PR TITLE
test: fix http2 frame metadata creation - cap payload size

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: fix http2 frame metadata creation in tests - cap payload size
Additional Description:
A wrong assert (`ASSERT(!payload_sequence->HasNext());`) is in the metadata frame creation which is triggered when the size of the map is too large.
This PR removes the assert, and adds comments to the function describing the cap.

Risk Level: low - tests only
Testing: N/A.
Docs Changes: N/A.
Release Notes: N/A.
Platform Specific Features: N/A.

### CODE REVIEW

**Comment by adisuissa** [comment]
Assigning to @birenroy as the previous updater of the function (#24946).

**Comment by birenroy** [comment]
Thanks for the PR!

File: test/common/http/http2/http2_frame.cc [hunk 1]
```diff
@@ -284,12 +285,12 @@ Http2Frame Http2Frame::makeMetadataFrameFromMetadataMap(uint32_t stream_index,
 
   auto payload_sequence = encoder.EncodeRepresentations(representations);
   ASSERT(payload_sequence->HasNext() || numberOfNameValuePairs == 0);
+  // Cap the payload size at 4MiB. Anything above that will be truncated.
   const size_t maxPayloadSize = 4 * 1024 * 1024;
   std::string payload;
   if (payload_sequence->HasNext()) {
```
[the remark this answers is not in the document]
Comment (adisuissa) [reply test/common/http/http2/http2_frame.cc]: Done.

File: test/common/http/http2/http2_frame.cc [hunk 2]
```diff
@@ -270,6 +270,7 @@ Http2Frame Http2Frame::makeWindowUpdateFrame(uint32_t stream_index, uint32_t inc
 }
 
 // Note: encoder in codebase persists multiple maps, with each map representing an individual frame.
+// The frame payload size is capped at 4MiB.
```
Comment (birenroy) [inline test/common/http/http2/http2_frame.cc:273]: This comment can be removed now.

File: test/common/http/http2/http2_frame.cc [hunk 3]
```diff
@@ -284,12 +285,12 @@ Http2Frame Http2Frame::makeMetadataFrameFromMetadataMap(uint32_t stream_index,
 
   auto payload_sequence = encoder.EncodeRepresentations(representations);
   ASSERT(payload_sequence->HasNext() || numberOfNameValuePairs == 0);
-  const size_t maxPayloadSize = 4 * 1024 * 1024;
-  std::string payload;
-  if (payload_sequence->HasNext()) {
-    payload = payload_sequence->Next(maxPayloadSize);
+  // Concatenate all the payload segments to a single string.
+  std::string payload("");
```
Comment (birenroy) [inline test/common/http/http2/http2_frame.cc:289]: I'm not sure the extra initialization parameter is useful. std::string is already initialized empty, right?

**Review by birenroy — APPROVED** [review]
Looks good, just two minor nits. Please also update the PR description.

File: test/common/http/http2/http2_frame.cc (hunk 2 above)
Comment (adisuissa) [reply test/common/http/http2/http2_frame.cc]: Done.

File: test/common/http/http2/http2_frame.cc (hunk 3 above)
Comment (adisuissa) [reply test/common/http/http2/http2_frame.cc]: Yeah, that's the same as the default c'tor initialization.
It's from an "old-school" view of c++ where it is better to have explicit initializations over leaving something that may be undefined.
Removed the explicit initialization to comply with newer style.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**